calliope/core.py

```diff
@@ -926,62 +907,60 @@ def read_data(self):
-        self._sets = {**self._sets, **ts_sets_constraints}
-        ts_sets_costs = {'y_def_' + k[1] + '_' + k[2]: set()
-                   for k in self.config_model.timeseries_data}
-        self._sets = {**self._sets, **ts_sets_costs}
+        ts_sets = {'y_def_' + param: set()
+                   for param in self.config_model.timeseries_constraints}
+        self._sets = {**self._sets, **ts_sets}
 
         for param in self.config_model.timeseries_constraints: #constraints
             param_data = {}
-            for y in self._sets['y']:
-                # First, set up each parameter without considering
-                # potential per-location (per-x) overrides
-                k = y + '.constraints.' + param
-                option = self.get_option(k)
-
-                df = self._read_param_for_tech(param, y, time_res, option, x=None)
-
-                for x in self._sets['x']:
-                    # Check for each x whether it defines an override
-                    # that is different from the generic option, and if so,
-                    # update the dataframe
-                    option_x = self.get_option(k, x=x)
-                    if option != option_x:
-                        df.loc[:, x] = self._read_param_for_tech(param, y, time_res, option_x, x=x)
-
-                self._validate_param_df(param, y, df)  # Have all `x` been set?
-
-                param_data[y] = xr.DataArray(df, dims=['t', 'x'])
+            if 'om' in param or 'sub' in param: # cost constraints
+                for y in self._sets['y']:
+                    df = {}
+                    for k in self._sets['k']:
+                        # First, set up each parameter without considering
+                        # potential per-location (per-x) overrides
+                        option = self.get_cost(param, y, k)
+
+                        df[k] = self._read_param_for_tech(param, y, time_res,
+                                                       option, x=None)
+                        for x in self._sets['x']:
+                            # Check for each x whether it defines an override
+                            # that is different from the generic option, and if so,
+                            # update the dataframe
+                            option_x = self.get_cost(param, y, k, x=x)
+                            if option != option_x:
+                                df[k].loc[:, x] = self._read_param_for_tech(
+                                                            param, y, time_res,
+                                                            option_x, x=x)
+
+                        self._validate_param_df(param, y, df[k])  # Have all `x` been set?
+                        # Create 3D dataframe with axes ordered as (t, x, k)
+                    df_k = pd.Panel(df).swapaxes(0,1).swapaxes(1,2)
+                    param_data[y] = xr.DataArray(df_k, dims=['t', 'x', 'k'])
```

Comment:
Happy to update it, haven't really spent time working with xarray DataArrays, so was falling back on my pandas knowledge there.